replay/models/cql.py

```diff
@@ -0,0 +1,354 @@
+        critic_encoder_factory: EncoderArg = "default",
+        q_func_factory: QFuncArg = "mean",
+        batch_size: int = 256,
+        n_frames: int = 1,
+        n_steps: int = 1,
+        gamma: float = 0.99,
+        tau: float = 0.005,
+        n_critics: int = 2,
+        initial_temperature: float = 1.0,
+        initial_alpha: float = 1.0,
+        alpha_threshold: float = 10.0,
+        conservative_weight: float = 5.0,
+        n_action_samples: int = 10,
+        soft_q_backup: bool = False,
+        use_gpu: UseGPUArg = False,
+        scaler: ScalerArg = None,
+        action_scaler: ActionScalerArg = None,
+        reward_scaler: RewardScalerArg = None,
+        **params
+    ):
+        super().__init__()
+        self.top_k = top_k
+        self.action_randomization_scale = action_randomization_scale
+        self.n_epochs = n_epochs
+        self.assert_omp_single_thread()
+
+        self.model = CQL_d3rlpy.CQL(
+            actor_learning_rate=actor_learning_rate,
+            critic_learning_rate=critic_learning_rate,
+            temp_learning_rate=temp_learning_rate,
+            alpha_learning_rate=alpha_learning_rate,
+            actor_optim_factory=actor_optim_factory,
+            critic_optim_factory=critic_optim_factory,
+            temp_optim_factory=temp_optim_factory,
+            alpha_optim_factory=alpha_optim_factory,
+            actor_encoder_factory=actor_encoder_factory,
+            critic_encoder_factory=critic_encoder_factory,
+            q_func_factory=q_func_factory,
+            batch_size=batch_size,
+            n_frames=n_frames,
+            n_steps=n_steps,
+            gamma=gamma,
+            tau=tau,
+            n_critics=n_critics,
+            initial_temperature=initial_temperature,
+            initial_alpha=initial_alpha,
+            alpha_threshold=alpha_threshold,
+            conservative_weight=conservative_weight,
+            n_action_samples=n_action_samples,
+            soft_q_backup=soft_q_backup,
+            use_gpu=use_gpu,
+            scaler=scaler,
+            action_scaler=action_scaler,
+            reward_scaler=reward_scaler,
+            **params
+        )
+
+    @staticmethod
+    def _rate_user_items(
+        model: bytes,
+        user_idx: int,
+        items: np.ndarray,
+    ) -> pd.DataFrame:
+        user_item_pairs = pd.DataFrame({
+            'user_idx': np.repeat(user_idx, len(items)),
+            'item_idx': items
+        })
+        input_batch = torch.from_numpy(
+            user_item_pairs.to_numpy()
+        ).float().cpu()
+
+        with io.BytesIO(model) as buffer:
+            model = torch.jit.load(buffer, map_location=torch.device('cpu'))
+
+        with torch.no_grad():
+            user_item_pairs['relevance'] = model.forward(input_batch).numpy()
+
+        # It doesn't explicitly filter seen items and doesn't return top k items.
+        # Instead, it returns all predictions as is to be filtered further by base methods
```

Comment:
Clean this comments, please

Reply:
moved this comment to `_predict` and made it clearer